### DISCUSSION TITLE
How can I control the `say()` function?

### DISCUSSION THREAD

**Discussion by shyagamzo** [johnlindquist/kit#1217] [Q&A] [answered]
I can run the `say()` function and it successfully reads my text.

A few questions:
1. Is there a way to stop it in the middle of reading longer texts?
2. How can I configure the speech engine?
    Shuffling through the codebase I say you add a second parameter with `rate` and `name`.
    What's the exact API for that parameter? It is currently typed as `any`.
3. What's the underlying engine used for TTS?

Thanks in advance!

Great work on ScriptKit! Cheers 🥂

**Comment by johnlindquist** [comment, marked as answer]
@shyagamzo 
The built-in Chrome engine. Here's the relevant source bits from the app:

```js
const synth = window.speechSynthesis;
if (synth.speaking) {
  synth.cancel();
}

const utterThis = new SpeechSynthesisUtterance(a?.text);
utterThis.rate = a?.rate || 1.3;
utterThis.pitch = a?.pitch || 1;
utterThis.lang = a?.lang || 'en-US';
const voices = synth.getVoices();
utterThis.voice =
  voices.find((v) => v.name === a?.name) || synth.getVoices()[0];
synth.speak(utterThis);
```

Here's a script showing "cancelling" `say` in various ways:

```js

/*
    You can use this block of code in any script to terminate a previous script.
*/
let scriptName = import.meta.url.split("/").pop().split(".").shift()

// Get the running process from the app
let processes = await getProcesses()
// match .js/.ts/.mjs endings
let scriptInstances = processes.filter(p => p.scriptPath.match(new RegExp(`${scriptName}\\.(js|ts|mjs)$`)))

if (scriptInstances.length > 1) {
  say("")
  // This will terminate both the previous script and this script
  scriptInstances.forEach((instance, index) => {
    process.kill(instance.pid)
  })
}

// Start saying long thing
say(`I have so much to say I'm just going to keep talking until someone shuts me up`)

registerShortcut("opt x", () => {
  say("") //will cancel
  process.exit() // you need to exit or else the shortcuts will keep the script active
})

registerShortcut("opt y", () => {
  say("You're done", {
    name: "Alice",
    rate: 0.5,
    pitch: 2,
  })
  process.exit()
})

// limitation: There's no "on complete" event for "say"
```

**Reply by shyagamzo** [reply]
Nice!! ✨
Works perfectly!

Anyway for delegating the `window` object?
I tried accessing it directly to get the supported voices and it crashes because it doesn't exist (which now makes sense 😅).

Edit:
I manually ran `window.speechSynthesis.getVoices()` in my devtools, so I now know what my options are.
Nevertheless, it would be nice being able to incorporate that into my script.

**Reply by johnlindquist** [reply]
You can access the app's dev tools here:

![CleanShot 2023-04-28 at 08 50 08](https://user-images.githubusercontent.com/36073/235180391-01188973-39c2-42f0-a41d-e74217a713bf.png)

I'll have to implement a way to "run js in window context" from the script. It's definitely doable, hadn't really thought about it until now 🤔
